paramiko/pkcs11.py

```diff
@@ -0,0 +1,160 @@
+from ctypes import (c_void_p, c_ulong, c_int, c_char_p, cast, addressof,
+                    sizeof, byref, cdll, Structure)
+import subprocess
+import os
+import errno
+from paramiko.ssh_exception import AuthenticationException
+
+
+class PKCS11Exception (Exception):
```

Comment:
I think we probably still want this to sub-class SSHException

Reply:
I was debating if I should.  Its technically separate from the SSH2 protocol, so I decided to keep the pkcs11 specific exceptions to not sub-class the SSHException.  For the PKCS11 authentication related exceptions (wrong pin, etc), I kept PKCS11AuthenticationException a subclass of AuthenticationException.

Reply:
That's a fair argument, which may prevail, but the counter-argument is that paramiko.SSHException is used for anything paramiko specific, not just SSH protocol specific. Various failures to read or parse rsa keys etc also raise "SSHException".

Reply:
Good point about other key type exceptions raising "SSHException".  I'll change it to sub-class SSHException.